src/stridelayout.jl

```diff
@@ -0,0 +1,269 @@
+
+"""
+If the contiguous dimension is not the dimension with `Stride_rank{1}`
+"""
+struct ContiguousBatch{N} end
+Base.@pure ContiguousBatch(N::Int) = ContiguousBatch{N}()
+_get(::ContiguousBatch{N}) where {N} = N
+
+"""
+contiguous_batch_size(::Type{T}) -> ContiguousBatch{N}
+
+Returns the size of contiguous batches if `!isone(stride_rank(T, contiguous_axis(T)))`.
+If `isone(stride_rank(T, contiguous_axis(T)))`, then it will return `ContiguousBatch{0}()`.
+If `contiguous_axis(T) == -1`, it will return `ContiguousBatch{-1}()`.
+If unknown, it will return `nothing`.
+"""
+contiguous_batch_size(x) = contiguous_batch_size(typeof(x))
+contiguous_batch_size(::Type) = nothing
+contiguous_batch_size(::Type{Array{T,N}}) where {T,N} = ContiguousBatch{0}()
+contiguous_batch_size(::Type{<:Tuple}) = ContiguousBatch{0}()
+contiguous_batch_size(::Type{<:Union{Transpose{T,A},Adjoint{T,A}}}) where {T,A<:AbstractVecOrMat{T}} = contiguous_batch_size(A)
+contiguous_batch_size(::Type{<:PermutedDimsArray{T,N,I1,I2,A}}) where {T,N,I1,I2,A<:AbstractArray{T,N}} = contiguous_batch_size(A)
+function contiguous_batch_size(::Type{S}) where {N,NP,T,A<:AbstractArray{T,NP},I,S <: SubArray{T,N,A,I}}
+    _contiguous_batch_size(S, contiguous_batch_size(A), contiguous_axis(A))
+end
+_contiguous_batch_size(::Any, ::Any, ::Any) = nothing
+@generated function _contiguous_batch_size(::Type{S}, ::ContiguousBatch{B}, ::Contiguous{C}) where {B,C,N,NP,T,A<:AbstractArray{T,NP},I,S <: SubArray{T,N,A,I}}
+    if I.parameters[C] <: AbstractUnitRange
+        Expr(:call, Expr(:curly, :ContiguousBatch, B))
+    else
+        Expr(:call, Expr(:curly, :ContiguousBatch, -1))
+    end
+end
+
+struct StrideRank{R} end
+Base.@pure StrideRank(R::NTuple{<:Any,Int}) = StrideRank{R}()
+_get(::StrideRank{R}) where {R} = R
+Base.collect(::StrideRank{R}) where {R} = collect(R)
+@inline Base.getindex(::StrideRank{R}, i::Integer) where {R} = R[i]
+@inline Base.getindex(::StrideRank{R}, ::Val{I}) where {R,I} = StrideRank{permute(R, I)}()
+
+function rank_to_sortperm(R::NTuple{N,Int}) where {N}
+    sp = ntuple(zero, Val{N}())
+    r = ntuple(n -> sum(R[n] .≥ R), Val{N}())
+    @inbounds for n in 1:N
+        sp = Base.setindex(sp, n, r[n])
+    end
+    sp
+end
+"""
+rank_to_sortperm(::StrideRank) -> NTuple{N,Int}
+
+Returns the `sortperm` of the stride ranks.
+"""
+@generated Base.sortperm(::StrideRank{R}) where {R} = rank_to_sortperm(R)
+
+stride_rank(x) = stride_rank(typeof(x))
+stride_rank(::Type) = nothing
+stride_rank(::Type{Array{T,N}}) where {T,N} = StrideRank{ntuple(identity, Val{N}())}()
+stride_rank(::Type{<:Tuple}) = StrideRank{(1,)}()
+
+stride_rank(::Type{B}) where {T, A, B <: Union{Transpose{T,A},Adjoint{T,A}}} = _stride_rank(B, stride_rank(A))
+_stride_rank(::Type{<:Union{Transpose{T,A},Adjoint{T,A}}}, ::Nothing) where {T,A<:AbstractMatrix{T}} = nothing
+_stride_rank(::Type{<:Union{Transpose{T,A},Adjoint{T,A}}}, rank) where {T,A<:AbstractMatrix{T}} = rank[Val{(2,1)}()]
```

Comment:
Transposes and adjoints of `AbstractVector`s don't seem to be handled by `stride_rank`?

Reply:
I'm thinking it should just return `(2,1)`?

Reply:
Yes, I think so as well.